onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorQLinearAveragePooling.cpp

```diff
@@ -0,0 +1,208 @@
+            std::vector<uint32_t> nchwInputStrides;
+            ConvertNHWCToNCHW(dimCount, inputSizes, nchwInputSizes, nchwInputStrides);
+            m_inputTensorDescs[OrtInputTensors::ortInput] = TensorDesc(m_inputTensorDescs[OrtInputTensors::ortInput].GetDmlDataType(), nchwInputSizes, nchwInputStrides);
+
+            gsl::span<const uint32_t> inputScaleSizes = m_inputTensorDescs[OrtInputTensors::ortInputScale].GetSizes();
+            std::vector<uint32_t> nchwInputScaleSizes;
+            std::vector<uint32_t> nchwInputScaleStrides;
+            ConvertNHWCToNCHW(dimCount, inputScaleSizes, nchwInputScaleSizes, nchwInputScaleStrides);
+            m_inputTensorDescs[OrtInputTensors::ortInputScale] = TensorDesc(m_inputTensorDescs[OrtInputTensors::ortInputScale].GetDmlDataType(),     nchwInputScaleSizes, nchwInputScaleStrides);
+
+            gsl::span<const uint32_t> inputZeroPointSizes = m_inputTensorDescs[OrtInputTensors::ortInputZeroPoint].GetSizes();
+            std::vector<uint32_t> nchwInputZeroPointSizes;
+            std::vector<uint32_t> nchwInputZeroPointStrides;
+            ConvertNHWCToNCHW(dimCount, inputZeroPointSizes, nchwInputZeroPointSizes, nchwInputZeroPointStrides);
+            m_inputTensorDescs[OrtInputTensors::ortInputZeroPoint] = TensorDesc(m_inputTensorDescs[OrtInputTensors::ortInputZeroPoint].GetDmlDataType(), nchwInputZeroPointSizes, nchwInputZeroPointStrides);
+
+            gsl::span<const uint32_t> outputScaleSizes = m_inputTensorDescs[OrtInputTensors::ortOutputScale].GetSizes();
+            std::vector<uint32_t> nchwOutputScaleSizes;
+            std::vector<uint32_t> nchwOutputScaleStrides;
+            ConvertNHWCToNCHW(dimCount, outputScaleSizes, nchwOutputScaleSizes, nchwOutputScaleStrides);
+            m_inputTensorDescs[OrtInputTensors::ortOutputScale] = TensorDesc(m_inputTensorDescs[OrtInputTensors::ortOutputScale].GetDmlDataType(), nchwOutputScaleSizes, nchwOutputScaleStrides);
+
+            gsl::span<const uint32_t> outputZeroPointSizes = m_inputTensorDescs[OrtInputTensors::ortOutputZeroPoint].GetSizes();
+            std::vector<uint32_t> nchwOutputZeroPointSizes;
+            std::vector<uint32_t> nchwOutputZeroPointStrides;
+            ConvertNHWCToNCHW(dimCount, outputZeroPointSizes, nchwOutputZeroPointSizes, nchwOutputZeroPointStrides);
+            m_inputTensorDescs[OrtInputTensors::ortOutputZeroPoint] = TensorDesc(m_inputTensorDescs[OrtInputTensors::ortOutputZeroPoint].GetDmlDataType(), nchwOutputZeroPointSizes, nchwOutputZeroPointStrides);
+
+            gsl::span<const uint32_t> outputSizes = m_outputTensorDescs[0].GetSizes();
+            std::vector<uint32_t> nchwOutputSizes;
+            std::vector<uint32_t> nchwOutputStrides;
+            ConvertNHWCToNCHW(dimCount, outputSizes, nchwOutputSizes, nchwOutputStrides);
+            m_outputTensorDescs[0] = TensorDesc(m_outputTensorDescs[0].GetDmlDataType(), nchwOutputSizes, nchwOutputStrides);
+        }
```

Comment:
This approach is allocation heavy and brittle (copy pasta typos are likely). We should be able to update the tensors in-place cleanly with a helper function that permutes the sizes and strides via a span of indices. I *thought* this helper function already existed on TensorDesc, but I'm not seeing it (must be somewhere else).

```
            ML_CHECK_VALID_ARGUMENT(dimensionCount >= 2);

            // Form a remapping for dimensions so C is moved before the spatial dimensions.
            // e.g. NWC   -> {0,2,1}     -> NCW
            //      NHWC  -> {0,3,1,2}   -> NCHW
            //      NDHWC -> {0,4,1,2,3} -> NCDHW
            std::vector<uint32_t> dimensionMapping = {0u, dimensionCount - 1u};
            dimensionMapping.resize(dimensionCount);
            std::iota(dimensionMapping.begin() + 2, dimensionMapping.end());

            m_inputTensorDescs[OrtInputTensors::ortInput].PermuteDimensions(dimensionMapping);
            m_inputTensorDescs[OrtInputTensors::ortInputScale].PermuteDimensions(dimensionMapping);
            m_inputTensorDescs[OrtInputTensors::ortInputZeroPoint].PermuteDimensions(dimensionMapping);
            m_inputTensorDescs[OrtInputTensors::ortOutputScale].PermuteDimensions(dimensionMapping);
            m_inputTensorDescs[OrtInputTensors::ortOutputZeroPoint].PermuteDimensions(dimensionMapping);
            m_outputTensorDescs[0].PermuteDimensions(dimensionMapping);

```

Reply:
We can even clean up all the statements above that reshape 1D to ND:

```
        if (kernelInfo.IsInputValid(OrtInputTensors::ortOutputZeroPoint))
        {
            m_inputTensorDescs[OrtInputTensors::ortOutputZeroPoint].PermuteDimensions(dimensionMapping, TensorAxis::RightAligned);
        }
```

e.g.
```
PermuteDimensions(...)
{
    EnsureStridesExist();
    EnsureDimensionCount(mapping.size(), alignment);
    PermuteArray(mapping);
}
```

Some related functions you can look at for inspiration include:
`DmlBufferTensorDesc::SetRankRightAligned`
`DmlBufferTensorDesc::ReorderDimensions`

Reply:
I am still working on this one

Reply:
Looking much nicer. Future generations reading this will thank you 😉.